src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+                hasSuggestedItemOptions = false;
+            }
+
+            var filterText = session.ApplicableToSpan.GetText(data.Snapshot);
+            var reason = data.Trigger.Reason;
+
+            // We do not care about the character in the case. We care about the reason only.
+            var roslynTrigger = Helpers.GetRoslynTrigger(data.Trigger, data.Trigger.Character);
+
+            if (!session.Properties.TryGetProperty(CompletionSource.InitialTriggerKind, out CompletionTriggerKind initialRoslynTriggerKind))
+            {
+                initialRoslynTriggerKind = CompletionTriggerKind.Invoke;
+            }
+
+            // Check if the user is typing a number. If so, only proceed if it's a number
+            // directly after a <dot>. That's because it is actually reasonable for completion
+            // to be brought up after a <dot> and for the user to want to filter completion
+            // items based on a number that exists in the name of the item. However, when
+            // we are not after a dot (i.e. we're being brought up after <space> is typed)
+            // then we don't want to filter things. Consider the user writing:
+            //
+            //      dim i =<space>
+            //
+            // We'll bring up the completion list here (as VB has completion on <space>).
+            // If the user then types '3', we don't want to match against Int32.
+            if (filterText.Length > 0 && char.IsNumber(filterText[0]))
+            {
+                if (!IsAfterDot(data.Snapshot, session.ApplicableToSpan))
+                {
+                    session.Dismiss();
+                    return null;
+                }
+            }
+
+            // We need to filter if a non-empty strict subset of filters are selected
+            var selectedFilters = data.SelectedFilters.Where(f => f.IsSelected).Select(f => f.Filter).ToImmutableArray();
+            var needToFilter = selectedFilters.Length > 0 && selectedFilters.Length < data.SelectedFilters.Length;
+            var filterReason = Helpers.GetFilterReason(roslynTrigger);
+
+            var initialListOfItemsToBeIncluded = new List<ExtendedFilterResult>();
+            foreach (var item in data.InitialSortedList)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (needToFilter && ShouldBeFilteredOutOfCompletionList(item, selectedFilters))
+                {
+                    continue;
+                }
+
+                if (!item.Properties.TryGetProperty(CompletionSource.RoslynItem, out RoslynCompletionItem roslynItem))
+                {
+                    roslynItem = RoslynCompletionItem.Create(
+                        displayText: item.DisplayText, 
+                        filterText: item.FilterText, 
+                        sortText: item.SortText,
+                        displayTextSuffix: item.Suffix);
+                }
```

Comment:
There are such scenarios when wee are called for filtering but didn't provide any item.

I explain my knowledge of this in other comment. @AmadeusW to provide more.

---
In reply to: [232509802](https://github.com/dotnet/roslyn/pull/29016#discussion_r232509802) [](ancestors = 232509802)

Reply:
Sorry, i apologize, but i do not remember seeing the comment.  Can you point me to it, or sum it up to jog my memory?  This seems super weird, and it's unclear to me why we would support this, versus just bailing immediately on these items.

Reply:
Sorry! It seems that there are similar questions and I'm responding them in one order and you're reading my responses in another order. Here is an explanation why we can be called for Filter even if the were not called or not provided items before: [https://github.com/dotnet/roslyn/pull/29016/files#diff-17beee9d668f1205a31fdf7e5a5dfa95R191](https://github.com/dotnet/roslyn/pull/29016/files#diff-17beee9d668f1205a31fdf7e5a5dfa95R191)

Copying it here just in case:
 CyrusNajmabadi a day ago  • 
 Member
how can we not have a completion service? #Resolved

  @ivanbasov
ivanbasov 6 hours ago  Member
Here is a scenario we just observed. In a Razor code, one types '@' and presses CTRL+J. Razor provides items to Editor. Roslyn either not called o provide items or responded that it could not provide any item.

Then, Editor decides to call Roslyn to filter items. All items were created by Razor. None of other has a reference to completion service.



---
In reply to: [232835880](https://github.com/dotnet/roslyn/pull/29016#discussion_r232835880) [](ancestors = 232835880)